baseplate/events/publisher.py

```diff
@@ -0,0 +1,203 @@
+# messages to batch
+MAX_BATCH_AGE = 1
+# seconds to wait between retries when the collector fails
+RETRY_DELAY = 2
+
+
+class Batcher(object):
+    """Time-aware batching producer.
+
+    A batch-consumer gives necessary context to this object. It must provide:
+        * batch_size_limit - the maximum size of an individual batch
+        * batch_size_overhead - base size cost of a batch
+        * get_item_size() - given an item, return its batch-relevant size
+        * consume_batch() - consume a fully formed batch
+
+    A flush may occur if the batch reaches the maximum specified size during
+    add() or if explicitly flush()ed.
+
+    """
+    def __init__(self, consumer):
+        self.consumer = consumer
+        self.batch = []
+        self.batch_size = self.consumer.batch_size_overhead
+        self.batch_start = None
+
+    @property
+    def batch_age(self):
+        """Return the age in seconds of the oldest item in the batch.
+
+        If there are no items in the batch, 0 is returned.
+
+        """
+        if not self.batch_start:
+            return 0
+        return time.time() - self.batch_start
+
+    def add(self, item):
+        """Add an item to the batch, potentially flushing."""
+        item_size = self.consumer.get_item_size(item)
+        if self.batch_size + item_size > self.consumer.batch_size_limit:
+            self.flush()
+        self.batch.append(item)
+        self.batch_size += item_size
+        if not self.batch_start:
+            self.batch_start = time.time()
+
+    def flush(self):
+        """Explicitly flush the batch if any items are enqueued."""
+        if self.batch:
+            self.consumer.consume_batch(self.batch)
+            self.batch = []
+            self.batch_size = self.consumer.batch_size_overhead
+            self.batch_start = None
+
+
+def gzip_compress(content):
+    buf = BytesIO()
+    with gzip.GzipFile(fileobj=buf, mode="wb", compresslevel=9) as f:
+        f.write(content)
+    return buf.getvalue()
+
+
+class BatchConsumer(object):
+    batch_size_overhead = 2  # the [] that wrap the json list
+    batch_size_limit = 500 * 1024
+
+    def __init__(self, metrics_client, cfg):
+        self.metrics = metrics_client
+        self.url = "https://%s/v1" % cfg.collector.hostname
+        self.key_name = cfg.key.name
+        self.key_secret = cfg.key.secret
+        self.session = requests.Session()
+
+    @staticmethod
+    def get_item_size(item):
+        # the item is already serialized, so we'll just add one byte for comma
+        return len(item) + 1
+
+    def _sign_payload(self, payload):
+        digest = hmac.new(self.key_secret, payload, hashlib.sha256).hexdigest()
+        return "key={key}, mac={mac}".format(key=self.key_name, mac=digest)
+
+    def consume_batch(self, items):
+        logger.info("sending batch of %d items", len(items))
+        payload = b"[" + b",".join(items) + b"]"
+        compressed_payload = gzip_compress(payload)
```

Comment:
I'm sure you already thought about this, but is there a chance that by always gzipping we're actually losing out and that we should consider not gzipping for items below a certain threshold? (I think we just did something with this somewhere else recently.) This talks about it a bit: http://webmasters.stackexchange.com/questions/31750/what-is-recommended-minimum-object-size-for-gzip-performance-benefits

Maybe we feel confident that even a single event would be big enough to always gzip?